ledger/catchpointlabel.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"encoding/base32"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+var base32Encoder = base32.StdEncoding.WithPadding(base32.NoPadding)
+
+// ErrCatchpointParsingFailed is used when we attempt to parse and catchpoint label and failing doing so.
+var ErrCatchpointParsingFailed = errors.New("catchpoing parsing failed")
+
+// CatchpointLabel represent a single catchpoint label. It will "assemble" a label based on the components
+type CatchpointLabel struct {
+	ledgerRound          basics.Round
+	ledgerRoundBlockHash crypto.Digest
+	balancesMerkleRoot   crypto.Digest
+	totals               AccountTotals
+}
+
+func makeCatchpointLabel(ledgerRound basics.Round, ledgerRoundBlockHash crypto.Digest, balancesMerkleRoot crypto.Digest, totals AccountTotals) CatchpointLabel {
+	return CatchpointLabel{
+		ledgerRound:          ledgerRound,
+		ledgerRoundBlockHash: ledgerRoundBlockHash,
+		balancesMerkleRoot:   balancesMerkleRoot,
+		totals:               totals,
+	}
+}
+
+// String return the user-facing representation of this catchpoint label. ( i.e. the "label" )
+func (l CatchpointLabel) String() string {
+	hash := l.Hash()
+	encodedHash := base32Encoder.EncodeToString(hash[:])
+	out := fmt.Sprintf("%d#%s", l.ledgerRound, encodedHash)
+	return out
+}
+
+// Hash return the hash portion of this catchpoint label
+func (l CatchpointLabel) Hash() crypto.Digest {
+	buffer := make([]byte, 2*crypto.DigestSize+256)
```

Comment:
Instead of 256, this should be `len(encodedTotals)`. Also, the `AccountTotals` struct lacks codec tags, so if we ever rename fields in that struct, this will break.

Reply:
I can't go all the way here with MicroAlgos codec-ing, as it would break the current AccountData.

Reply:
done.